fetch.py

```diff
@@ -0,0 +1,152 @@
+from tqdm import tqdm
+"""
+Decorator for downloading or updating files required for a function.
+"""
+def reqfile(func=None, path=None, url=None):
+    if not func:
+        return functools.partial(fetch_factory, path=path, url=url)
+    @functools.wraps(test_func)
+    def wrapper(*args, **kwargs):
+        fetch(path, url)
+        return func(*args, **kwargs)
+    return wrapper
+
+"""
+Downloads a file from url to a given path.
+Also shows a progress bar to keep track.
+"""
+def downfile(path, url):
+    r = requests.get(url, stream=True)
+    r.raise_for_status()
+    total_size = int(r.headers.get('content-length', 0))
+    with open(path, 'wb') as f:
+        chunk_size = 64*1024
+        pbar = tqdm(total=total_size, unit='B', unit_scale=True)
+        for data in r.iter_content(chunk_size):
+            f.write(data)
+            pbar.update(chunk_size)
+
+"""
+Checks a given path for file and downloads if a url is given or file name
+is an accession id. Also downloads the file if the remote location has a
+more recent version. Gives an IOError if path does not exist and an accession id
+can not be found.
+    Usage:
+        fetch('NC_000853.txt')
+        fetch('DRDatabase.txt', 'http://crispr.i2bc.paris-saclay.fr/crispr/BLAST/DR/DRdatabase')
+    Output:
+    NC_000853.txt file would be created in the running folder, if it existed before,
+    it would be updated to a newer version if available in Entrez database. DRDatabase
+    would also be created in the running folder, if the url has a more recent version
+    the local file would be updated.
+"""
+def fetch(path=None, url=None):
+    """
+    Checks and downloads a url if last modified date does not exist for
+    the url or is more recent than local file.
+    """
+    def sync():
+        if not path_exists:
+            downfile(path, url)
+            return
+        '''
+        Check last modified dates of file and url, download if url is newer.
+        '''
+        filemodtime = datetime.datetime.fromtimestamp(os.path.getmtime(path))
+        r = requests.get(url, stream=True)
+        if 'Last-Modified' not in r.headers.keys():
+            print('Last modified header not found in url, downloading...')
+            downfile(path, url) #no last-modified header in url, downloading file
+        urlmodstr = r.headers['Last-Modified']
+        urlmodtime = datetime.datetime.strptime(urlmodstr, '%a, %d %b %Y %H:%M:%S %Z')
+        if filemodtime < urlmodtime:
+            '''
+            Url file is more recent, downloading url.'''
+            print('Url is more recent than file, downloading')
+            downfile(path, url)
+            return
+        print('File is recent, returning')
+        return
+    """
+    Checks and downloads an accession file from nuccore database if file does
+    not exist or its dates are different from url version.
+    """
+    def gbsync():
+        print('Trying to fetch from Entrez')
+        regex = r'(\w{2}_\d{6}|\w{1}_\d{5})'
```

Comment:
Some accession numbers are longer than 6 numbers after the underscore, and sometimes it's a mix of letters and numbers.

These are all valid accessions, for example:

```
NZ_CP008702
NZ_AHJQ01000001
```